### ISSUE TITLE
bug: OWLv2 subprocess stalls at ~63 inferences, 90s watchdog restart creates detection blind window

### ISSUE THREAD

**Issue by thaije** [thaije/derpbot-explorer#34]
## Observed behaviour

During the benchmark smoke test (easy, seed=1, speed=2, 900s), the OWLv2 detector subprocess ran ~63 inferences then stalled — no further results were produced. The 90s watchdog in `_relay_loop` detected the stall and restarted the subprocess. This pattern repeated throughout the run.

Log signature:
```
Detector: relay alive, 63 inferences so far, subprocess alive=True
# ... then silence for ~90 wall-seconds ...
Detector: no inference results for 90s — subprocess hung, restarting. (frame_q_empty=False)
Detector: new inference subprocess started — waiting for ready.
```

## Impact

Each stall = ~90s blind window where no detections are produced. OWLv2 warm-up takes additional time after each restart. In a 900s run with multiple stall-restart cycles, a significant portion of exploration time produces zero detections. Likely contributes to missed objects (e.g. 2 missed in the smoke test run, 2 FPs possibly from a post-restart confidence miscalibration).

## Run context

- easy_900s, seed=1, speed=2
- 4/6 objects found (fire_ext ×2, first_aid_kit ×1, person ×1)
- Subprocess first signalled ready normally, then stalled at inference ~63

## Hypothesis

CUDA memory accumulation in the subprocess. Each inference allocates GPU tensors (`inputs`, `outputs`, `target_sizes`) on `cuda:0`. If these are not being released (e.g. graph retained by autocast or a reference leak), VRAM fills up after ~60 forward passes and the next `model(**_inputs)` hangs or OOMs silently. The subprocess process then stops consuming frames but stays alive (no Python exception propagates out).

GPU 0 is shared with Gazebo's renderer, so available VRAM is reduced and the ceiling is lower than a dedicated-inference setup.

## Possible fixes

1. **Explicit tensor deletion after each inference** — `del inputs, outputs, target_sizes; torch.cuda.empty_cache()` at the end of each forward pass in `_inference_worker`.
2. **Limit VRAM via `torch.cuda.set_per_process_memory_fraction`** — caps the subprocess's VRAM budget so an OOM raises immediately (catchable) rather than causing a hang.
3. **`torch.no_grad()` + `torch.inference_mode()` context** — already uses `no_grad`, but `inference_mode` is stricter and releases more intermediate state.
4. **Subprocess health check via `_worker.is_alive()` in the stall watchdog** — currently the watchdog only triggers if `_worker.is_alive()` is True; if the process zombie-hangs with alive=True but no output, the 90s timeout fires correctly. Verify the subprocess actually exits after each restart (no zombie accumulation).

## Related

- #26 — silent detection failures (0 detections with alive=True); the stall at ~63 inferences may be the same root cause — if prior code lacked the restart path or it failed silently, the run would log inferences up to the stall point then produce nothing afterward.
- `agent/detector.py` — `_relay_loop` stall watchdog at line 385; `_inference_worker` tensor lifecycle lines 161–181.

## Reproducer

Run any 600–900s scenario with full perception. Watch for `subprocess alive=True` in the 30s heartbeat log followed by silence and then the restart message.

**Comment by thaije** [comment]
## Smoke test verification — seed=1, easy, speed=2

Tested the `del outputs, inputs, target_sizes` + `torch.cuda.empty_cache()` every 20 inferences fix, and the switch from `torch.no_grad()` to `torch.inference_mode()`.

**Result: stall delayed but NOT fixed.**

| Metric | Previous run | This run (seed=1) |
|--------|-------------|-------------------|
| Stall at inference # | ~63 | ~84 |
| Post-stall behaviour | Restart, re-stall | Restart loop (5+ attempts, 0 new inferences each) |
| Overall score | 59.5 C | 72.5 B |
| found_ratio | 4/6 (0.67) | 6/6 (1.0) |
| Coverage | — | 59.6% |
| Collisions | 0 | 0 |

### What happened

- Inferences grew normally through #84, then stalled.
- First stall triggered at ~84 (vs ~63 before). The `del` + `empty_cache()` postponed it by ~21 inferences.
- **Every subsequent restart produced 0 inferences.** The watchdog restarted the subprocess, it connected (`subprocess alive=True`), but never processed a single frame. This repeated 5+ times with `frame_q_empty=True` (no frames queued — the robot had stopped moving, so no new images to process).
- All 6 targets were found before the stall, so score recovered despite the blind second half.

### Root cause still unresolved

The `del` + `empty_cache()` mitigates gradual VRAM accumulation but doesn't fix the underlying leak/hang. The restart loop is a new concern: once stalled, even fresh subprocesses can't recover (possibly the GPU state is poisoned for the rest of the run, or frame_q_empty means the robot stopped exploring and thus no new images arrive to trigger inference).

### Next steps to investigate

1. Check whether `frame_q_empty=True` means the robot stopped moving (frontier exhaustion) rather than a subprocess bug — if so, this is expected behaviour in late-run patrol mode.
2. Profile GPU memory at inference #60, #70, #80 to see if VRAM still climbs despite `del` + `empty_cache()`.
3. Consider `torch.cuda.set_per_process_memory_fraction()` to force an audible OOM instead of a silent hang.
4. Test with a CUDA memory snapshot (`torch.cuda.memory._snapshot()`) before/after the stall to identify what's leaking.

**Comment by thaije** [comment]
## Root cause identified and fixed

### Root cause: CUDA memory leak via `results` dict

The `post_process_grounded_object_detection()` return value is a `dict` with GPU tensors (`scores`, `labels`, `boxes`). The old code only deleted `outputs`, `inputs`, `target_sizes` — but `results` held sliced references derived from `outputs` that kept CUDA memory alive. Over ~60-80 inferences, VRAM accumulated and the subprocess hung on the next allocation.

### Fix (3 parts in `agent/detector.py`)

1. **CPU tensor conversion**: `.cpu().tolist()` on `results["scores"]`, `results["labels"]`, `results["boxes"]` immediately after post-processing. This releases GPU references so the CUDA allocator can reclaim memory.

2. **`torch.cuda.empty_cache()` every iteration** (instead of every 20). `empty_cache()` costs ~0.5ms — negligible at 5 Hz. The previous every-20 approach let the caching allocator fragment between flushes. Per-invalidation flush prevents the ~80-inference accumulation ceiling.

3. **`results` added to `del` statement**: `finally: del outputs, inputs, target_sizes, results` ensures no GPU tensor survives past the cleanup block.

### VRAM telemetry confirms the fix

```
inf#10  alloc=309MB reserved=492MB
inf#20  alloc=309MB reserved=492MB
inf#30  alloc=309MB reserved=492MB
inf#40  alloc=309MB reserved=492MB
...
inf#90  alloc=309MB reserved=492MB
```

Flat from start to finish. Previous runs leaked until hang at ~63–84 inferences.

### Restart-loop fix (separate issue)

The watchdog was restarting the subprocess when `frame_q_empty=True` — meaning the robot was stationary/patrolling and no camera frames were queued. This created a false restart loop. Fixed: skip restart when `frame_q_empty`, log as `"subprocess idle, not hung"`, and reset the watchdog timer so it doesn't keep triggering.

### Score comparison (seed=1, easy, speed=2)

| Metric | Previous (broken) | This fix |
|--------|-------------------|----------|
| Subprocess stall? | Yes, at #63-84, restart loop | **No** — 98 inferences, no stall |
| VRAM trend | Leaking → hang | **Flat 309MB** |
| False restarts | 5+ | **0** |
| Overall score | 59.5 C / 72.5 B* | 68.0 C |
| found_ratio | 0.67 / 1.0* | 0.83 |
| Coverage | — / 59.6%* | 72.2% |

*72.5 B run was a different seed=1 run where all 6 objects were found before stall hit.

### Remaining observations

- `_image_cb` diagnostic logging added (every 500th raw callback). Camera does keep publishing throughout the run.
- 98 inferences in ~6 min of active exploration, then ~3 min of idle (patrol/stuck nav) — the `frame_q_empty` idle detection is correct.
- Seed variance still affects score significantly. The detector fix is verified; remaining score improvements come from nav/exploration improvements (#8, #17).

**Comment by thaije** [comment]
Fixed in commit 6de18a8. Root cause was  dict holding GPU tensors after , plus infrequent . Fix: CPU tensor conversion (),  in finally block, per-iteration , and idle-detection for the watchdog. Verified: VRAM flat at 309MB, 98 inferences with zero stalls.

**Comment by thaije** [comment]
Fixed in commit 6de18a8.

Root cause: post_process_grounded_object_detection() returns GPU tensors (scores/labels/boxes) that were never released. Added CPU conversion (.cpu().tolist()), del results in finally block, per-iteration torch.cuda.empty_cache(), and idle-detection for the watchdog.

Verified: VRAM flat at 309MB/492MB across 98 inferences, zero stalls, zero false restarts.
